`unused_parens` lint suggestion also removes attributes on inner expression with `feature(stmt_expr_attributes)`.

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#![feature(stmt_expr_attributes)]
#![allow(unused_must_use)]

pub fn foo() -> impl Fn() {
    let _ = (#[inline] || println!("Hello!"));
    (#[inline] || println!("Hello!"));
    (#[inline] || println!("Hello!"))
}
```

I expected to see this happen: The `unused_parens` warning fires on all three lines of `foo`, and the suggestions should only suggest removing the parentheses, not the attribute also.

Instead, this happened: The `unused_parens` fires on the first and last lines of `foo`, and the suggestion also removes the `#[inline]`.
No `unused_parens` warning fires on the second line `(#[inline] || println!("Hello!"));` (but this is probably a separate issue, unrelated to attributes; it happens even when the attribute is removed; maybe `unused_must_use` overrides it or something).

```rs
warning: unnecessary parentheses around assigned value
 --> a.rs:5:13
  |
5 |     let _ = (#[inline] || println!("Hello!"));
  |             ^^^^^^^^^^^                     ^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
5 -     let _ = (#[inline] || println!("Hello!"));
5 +     let _ = || println!("Hello!");
  |

warning: unnecessary parentheses around block return value
 --> a.rs:7:5
  |
7 |     (#[inline] || println!("Hello!"))
  |     ^^^^^^^^^^^                     ^
  |
help: remove these parentheses
  |
7 -     (#[inline] || println!("Hello!"))
7 +     || println!("Hello!")
  |

warning: 2 warnings emitted
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.82.0-nightly (0d634185d 2024-08-29)
binary: rustc
commit-hash: 0d634185dfddefe09047881175f35c65d68dcff1
commit-date: 2024-08-29
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0
```

Requires use of `#![feature(stmt_expr_attributes)]`, otherwise the `#[inline]` is not legal in that position (though note that the `unused_parens` warning and suggestion is the same on stable, there's just an `error[E0658]: attributes on expressions are experimental` first).

The issue is not specific to closures and `#[inline]`, but that was the simplest example of a reasonable attribute-on-expression I thought of.

@rustbot label +A-diagnostics +requires-nightly

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"surechen"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_lint/src/unused.rs
+++ b/compiler/rustc_lint/src/unused.rs
@@ -804,7 +804,15 @@ trait UnusedDelimLint {
                 .find_ancestor_inside(value.span)
                 .map(|span| (value.span.with_hi(span.lo()), value.span.with_lo(span.hi()))),
             ast::ExprKind::Paren(ref expr) => {
-                expr.span.find_ancestor_inside(value.span).map(|expr_span| {
+                // For the expr with attributes, like `let _ = (#[inline] || println!("Hello!"));`,
+                // the span should contains the attributes, or the suggestion will remove them.
+                let expr_span_with_attrs =
+                    if let Some(attr_lo) = expr.attrs.iter().map(|attr| attr.span.lo()).min() {
+                        expr.span.with_lo(attr_lo)
+                    } else {
+                        expr.span
+                    };
+                expr_span_with_attrs.find_ancestor_inside(value.span).map(|expr_span| {
                     (value.span.with_hi(expr_span.lo()), value.span.with_lo(expr_span.hi()))
                 })
             }

--- a/tests/ui/lint/unused/unused-parens-for-stmt-expr-attributes-issue-129833.fixed
+++ b/tests/ui/lint/unused/unused-parens-for-stmt-expr-attributes-issue-129833.fixed
@@ -1,0 +1,15 @@
+//@ run-rustfix
+// Check the `unused_parens` suggestion for paren_expr with attributes.
+// The suggestion should retain attributes in the front.
+
+#![feature(stmt_expr_attributes)]
+#![deny(unused_parens)]
+
+pub fn foo() -> impl Fn() {
+    let _ = #[inline] #[allow(dead_code)] || println!("Hello!"); //~ERROR unnecessary parentheses
+    #[inline] #[allow(dead_code)] || println!("Hello!") //~ERROR unnecessary parentheses
+}
+
+fn main() {
+    let _ = foo();
+}

--- a/tests/ui/lint/unused/unused-parens-for-stmt-expr-attributes-issue-129833.rs
+++ b/tests/ui/lint/unused/unused-parens-for-stmt-expr-attributes-issue-129833.rs
@@ -1,0 +1,15 @@
+//@ run-rustfix
+// Check the `unused_parens` suggestion for paren_expr with attributes.
+// The suggestion should retain attributes in the front.
+
+#![feature(stmt_expr_attributes)]
+#![deny(unused_parens)]
+
+pub fn foo() -> impl Fn() {
+    let _ = (#[inline] #[allow(dead_code)] || println!("Hello!")); //~ERROR unnecessary parentheses
+    (#[inline] #[allow(dead_code)] || println!("Hello!")) //~ERROR unnecessary parentheses
+}
+
+fn main() {
+    let _ = foo();
+}

--- a/tests/ui/lint/unused/unused-parens-for-stmt-expr-attributes-issue-129833.stderr
+++ b/tests/ui/lint/unused/unused-parens-for-stmt-expr-attributes-issue-129833.stderr
@@ -1,0 +1,31 @@
+error: unnecessary parentheses around assigned value
+  --> $DIR/unused-parens-for-stmt-expr-attributes-issue-129833.rs:9:13
+   |
+LL |     let _ = (#[inline] #[allow(dead_code)] || println!("Hello!"));
+   |             ^                                                   ^
+   |
+note: the lint level is defined here
+  --> $DIR/unused-parens-for-stmt-expr-attributes-issue-129833.rs:6:9
+   |
+LL | #![deny(unused_parens)]
+   |         ^^^^^^^^^^^^^
+help: remove these parentheses
+   |
+LL -     let _ = (#[inline] #[allow(dead_code)] || println!("Hello!"));
+LL +     let _ = #[inline] #[allow(dead_code)] || println!("Hello!");
+   |
+
+error: unnecessary parentheses around block return value
+  --> $DIR/unused-parens-for-stmt-expr-attributes-issue-129833.rs:10:5
+   |
+LL |     (#[inline] #[allow(dead_code)] || println!("Hello!"))
+   |     ^                                                   ^
+   |
+help: remove these parentheses
+   |
+LL -     (#[inline] #[allow(dead_code)] || println!("Hello!"))
+LL +     #[inline] #[allow(dead_code)] || println!("Hello!")
+   |
+
+error: aborting due to 2 previous errors
+

